// @sudoku/stores/branchPoints.js
import { writable, get } from 'svelte/store';
/**
 * step is the index of history.
 */
function createBranchPoints() {
    const branchPoints = writable([]);

    return {
        subscribe: branchPoints.subscribe,
        add: (step) => {
            branchPoints.update(steps => [...steps, step]);
        },
        remove: (step) => {
            branchPoints.update(steps => steps.filter(s => s !== step));
        },
        update: branchPoints.update
            
    }
}

export const branchPoints = createBranchPoints();

function createHistory() {
    const history = writable([]);

    return {
        subscribe: history.subscribe,
        add: ({x,y}) => {
            history.update(cursors => [...cursors, {x , y } ]);
        },
        remove: ({x,y}) => {
            // 每次remove都要检查branchPoints是否有不存在的step；
            history.update(cursors => cursors.filter(c => c.x !== x || c.y !== y));
            branchPoints.update(steps => steps.filter(s => s < get(history).length));
        }
    }
}

export const history = createHistory();

function createRightNumber() {
    const rightNumber = writable(0);

    return {
        subscribe: rightNumber.subscribe,
        set: rightNumber.set
    }
}

export const rightNumber = createRightNumber();

function createWrongNumber() {
    const wrongNumber = writable(0);

    return {
        subscribe: wrongNumber.subscribe,   
        set: wrongNumber.set
    }
}

export const wrongNumber = createWrongNumber();
